pkg/domain/blueprintSpec.go

```diff
@@ -0,0 +1,219 @@
+	Id                   string
+	Blueprint            Blueprint
+	BlueprintMask        BlueprintMask
+	EffectiveBlueprint   EffectiveBlueprint
+	StateDiff            StateDiff
+	BlueprintUpgradePlan BlueprintUpgradePlan
+	Config               BlueprintConfiguration
+	Status               StatusPhase
+	// PersistenceContext can hold generic values needed for persistence with repositories, e.g. version counters or transaction contexts.
+	// This field has a generic map type as the values within it highly depend on the used type of repository.
+	// This field should be ignored in the whole domain.
+	PersistenceContext map[string]interface{}
+	Events             []interface{}
+}
+
+type StatusPhase string
+
+const (
+	// StatusPhaseNew marks a newly created blueprint-CR.
+	StatusPhaseNew StatusPhase = ""
+	// StatusPhaseStaticallyValidated marks the given blueprint spec as validated.
+	StatusPhaseStaticallyValidated StatusPhase = "staticallyValidated"
+	// StatusPhaseValidated marks the given blueprint spec as validated.
+	StatusPhaseValidated StatusPhase = "validated"
+	// StatusPhaseInvalid marks the given blueprint spec is semantically incorrect.
+	StatusPhaseInvalid StatusPhase = "invalid"
+	// StatusPhaseInProgress marks that the blueprint is currently being processed.
+	StatusPhaseInProgress StatusPhase = "inProgress"
+	// StatusPhaseFailed marks that an error occurred during processing of the blueprint.
+	StatusPhaseFailed StatusPhase = "failed"
+	// StatusPhaseCompleted marks the blueprint as successfully applied.
+	StatusPhaseCompleted StatusPhase = "completed"
+)
+
+type BlueprintConfiguration struct {
+	// Force blueprint upgrade even when a dogu is unhealthy
+	IgnoreDoguHealth bool
+	// allowNamespaceSwitch allows the blueprint upgrade to switch a dogus namespace
+	AllowDoguNamespaceSwitch bool
+}
+
+type StateDiff struct{}
+
+type BlueprintUpgradePlan struct {
+	DogusToInstall   []string
+	DogusToUpgrade   []string
+	DogusToUninstall []string
+
+	ComponentsToInstall   []string
+	ComponentsToUpgrade   []string
+	ComponentsToUninstall []string
+
+	RegistryConfigToAdd    []string
+	RegistryConfigToUpdate []string
+	RegistryConfigToRemove []string
+}
+
+type BlueprintSpecInvalidEvent struct {
+	ValidationError error
+}
+
+type BlueprintSpecValidatedEvent struct{}
+
+type EffectiveBlueprintCalculatedEvent struct {
+	EffectiveBlueprint EffectiveBlueprint
+}
+
+// ValidateStatically checks the blueprintSpec for semantic errors and sets the status to the result.
+// Here will be only checked, what can be checked without any external information, e.g. without dogu specification.
+// returns a domain.InvalidBlueprintError if blueprint is invalid
+// or nil otherwise.
+func (spec *BlueprintSpec) ValidateStatically() error {
+	switch spec.Status {
+	case StatusPhaseNew: //continue
+	case StatusPhaseInvalid: //do not validate again
+		return &InvalidBlueprintError{Message: "blueprint spec was marked invalid before. Do not revalidate"}
```

Comment:
```suggestion
		return &InvalidBlueprintError{Message: "blueprint spec was marked invalid before: do not revalidate"}
```

Error messages in golang should be separated by colons and start in lowercase.